src/main/java/com/rackspace/jenkins_nodepool/NodePool.java

```diff
@@ -543,30 +581,49 @@ void attemptProvisionNode2(final NodeRequest request, final int timeoutInSec) th
                 // it returns before the launch is complete, so errors are
                 // not handled.
                 final NodePoolSlave nps = new NodePoolSlave(node, getCredentialsId(), nodePoolJob);
+                nodePoolJob.setNodePoolSlave(nps);
                 Jenkins.getInstance().addNode(nps);
 
-                LocalDateTime launchDeadline = LocalDateTime.now().plusMinutes(LAUNCH_TIMEOUT_MINUTES);
+                LocalDateTime launchDeadline = LocalDateTime.now().plusSeconds(installTimeoutSec);
                 NodePoolComputer npc;
                 while (true){
                     npc = (NodePoolComputer) Jenkins.getInstance().getComputer(nodePoolJob.getNodePoolNode().getName());
-                    if(LocalDateTime.now().isAfter(launchDeadline)){
+
+                    if(npc != null && !npc.isOffline()){
+                        // node is online, great, stop waiting for it.
+                        break;
+                    }
+
+                    // Check for conditions that mean that we no longer
+                    // need to wait for this node (Build finished and Timeout)
+                    if (!nodePoolJob.getRun().isBuilding()){
+                        // If the build has completed, we no longer care if
+                        // the node managed to come online, so stop waiting.
                         break;
                     }
-                    if(npc == null|| npc.isOffline()){
-                        Thread.sleep(10000);
-                        continue;
+                    if(LocalDateTime.now().isAfter(launchDeadline)){
+                        LOG.warning("Launch deadline expired: "+nps.getDisplayName()+" For: "+nodePoolJob.getRun().getDisplayName());
+                        break;
                     }
+
+                    Thread.sleep(500);
                 }
-                if (npc == null || npc.isOffline()){
-                    throw new Exception("Failed to launch Jenkins agent on "+nps.getNodePoolNode().getName());
+
+                if (nodePoolJob.getRun().isBuilding()){
+                    // build still running
+                    if( npc == null || npc.isOffline()){
+                        // build still running and node failed to come online
+                        throw new NodePoolException("Failed to launch Jenkins agent on "+nps.getNodePoolNode().getName()+" NPC: "+npc);
+                    }else{
```

Comment:
It'd be super-nice if spacing was consistent. There appear to be instances of ``}else{`` and ``} else {``, and similar spacing differences elsewhere, like line 612 above.

Reply:
There are many java style guides, we don't test against any of them. If we introduced one it would require a large cleanup, of our code and code we've vendored, so definitely another card.